pkg/util/cache/lru_cache.go

```diff
@@ -0,0 +1,164 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package cache
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
+	"unsafe"
+)
+
+var (
+	// There are multiple instances of the interner, one per worker. Counters are normally fine,
+	// gauges require special care to make sense. We don't need to clean up when an instance is
+	// dropped, because it only happens on agent shutdown.
+	tlmSIResets = telemetry.NewSimpleCounter("dogstatsd", "string_interner_resets",
+		"Amount of resets of the string interner used in dogstatsd")
+	tlmSIRSize = telemetry.NewSimpleGauge("dogstatsd", "string_interner_entries",
+		"Number of entries in the string interner")
+	tlmSIRBytes = telemetry.NewSimpleGauge("dogstatsd", "string_interner_bytes",
+		"Number of bytes stored in the string interner")
+	tlmSIRHits = telemetry.NewSimpleCounter("dogstatsd", "string_interner_hits",
+		"Number of times string interner returned an existing string")
+	tlmSIRMiss = telemetry.NewSimpleCounter("dogstatsd", "string_interner_miss",
+		"Number of times string interner created a new string object")
+	tlmSIRNew = telemetry.NewSimpleCounter("dogstatsd", "string_interner_new",
+		"Number of times string interner was created")
+	tlmSIRStrBytes = telemetry.NewSimpleHistogram("dogstatsd", "string_interner_str_bytes",
+		"Number of times string with specific length were added",
+		[]float64{1, 2, 4, 8, 16, 32, 64, 128})
+)
+
+// stringInterner is a string cache providing a longer life for strings,
+// helping to avoid GC runs because they're re-used many times instead of
+// created every time.
+type stringCacheItem struct {
+	s string // points into the mmap region
+	// LRU doubly-linked list.  Head is most recently used, tail is
+	// next candidate for eviction.
+	prev, next *stringCacheItem
+}
+
+type lruStringCache struct {
+	// LRU cache in strings.
+	strings map[string]*stringCacheItem
+	// LRU linked list head/tail.
+	head, tail *stringCacheItem
+	maxSize    int
+	curBytes   int
+	tlmEnabled bool
+}
+
+func newLruStringCache(maxSize int, tlmEnabled bool) lruStringCache {
+	if tlmEnabled {
+		tlmSIRNew.Inc()
+	}
+	return lruStringCache{
+		strings:    make(map[string]*stringCacheItem),
+		maxSize:    maxSize,
+		tlmEnabled: tlmEnabled,
+	}
+}
+
+// lookupOrInsert looks inside the cache for this key.  If found, it makes it the head of the LRU cache.
+// if not, it may evict the least recently used entry to make space.  It uses the allocator arg function
+// to create the string, in case there's a separate backing store for it.
+func (c *lruStringCache) lookupOrInsert(key []byte, allocator func(key []byte) string) string {
```

Comment:
Just as a general comment, it would help with readability of this function if the `deleteNode` and `insertNode` functionality were factored out into their own methods. This method got quite long here with the in-line implementation of inserting and deleting.

Reply:
I'll extract out the parts.